clustered/common/src/main/java/org/ehcache/clustered/common/messages/ServerStoreOpCodec.java

```diff
@@ -15,74 +15,63 @@
+class ServerStoreOpCodec {
+
 
-  private static final Charset UTF_8 = Charset.forName("UTF-8");
+  private static final byte STORE_OP_CODE_SIZE = 1;
+  private static final byte CACHE_ID_LEN_SIZE = 4;
+  private static final byte KEY_SIZE = 8;
+  private static final byte CHAIN_LEN_SIZE = 4;
 
-  private static final byte MSG_TYPE_OFFSET = 1;
-  private static final byte STORE_OP_CODE_OFFSET = 1;
-  private static final byte CACHE_ID_LEN_OFFSET = 4;
-  private static final byte KEY_OFFSET = 8;
-  private static final byte CHAIN_LEN_OFFSET = 4;
+  private final ChainCodec chainCodec;
 
-  public static byte[] encode(ServerStoreOpMessage message) {
+  ServerStoreOpCodec() {
+    this.chainCodec = new ChainCodec();
+  }
+
+  public byte[] encode(ServerStoreOpMessage message) {
     // TODO: improve data send over n/w by optimizing cache Id
+    ByteBuffer encodedMsg;
     int cacheIdLen = message.getCacheId().length();
     switch (message.operation()) {
       case GET:
-        ByteBuffer encodedMsg = ByteBuffer.allocate(MSG_TYPE_OFFSET + STORE_OP_CODE_OFFSET + CACHE_ID_LEN_OFFSET + KEY_OFFSET + cacheIdLen);
-        encodedMsg.put(EhcacheEntityMessage.Type.SERVER_STORE_OP.getOpCode());
-        encodedMsg.putInt(cacheIdLen);
-        encodedMsg.put(message.getCacheId().getBytes(UTF_8));
-        encodedMsg.putLong(message.getKey());
-        encodedMsg.put(message.operation().getStoreOpCode());
+        encodedMsg = ByteBuffer.allocate(STORE_OP_CODE_SIZE + CACHE_ID_LEN_SIZE + KEY_SIZE + 2 * cacheIdLen);
+        putCacheIdKeyAndOpCode(encodedMsg, message.getCacheId(), message.getKey(), message.operation().getStoreOpCode());
         return encodedMsg.array();
       case APPEND:
         AppendMessage appendMessage = (AppendMessage)message;
-        encodedMsg = ByteBuffer.allocate(MSG_TYPE_OFFSET + STORE_OP_CODE_OFFSET + CACHE_ID_LEN_OFFSET + KEY_OFFSET + cacheIdLen + appendMessage
+        encodedMsg = ByteBuffer.allocate(STORE_OP_CODE_SIZE + CACHE_ID_LEN_SIZE + KEY_SIZE + 2 * cacheIdLen + appendMessage
             .getPayload()
             .remaining());
-        encodedMsg.put(EhcacheEntityMessage.Type.SERVER_STORE_OP.getOpCode());
-        encodedMsg.putInt(cacheIdLen);
-        encodedMsg.put(message.getCacheId().getBytes(UTF_8));
-        encodedMsg.putLong(message.getKey());
-        encodedMsg.put(message.operation().getStoreOpCode());
+        putCacheIdKeyAndOpCode(encodedMsg, message.getCacheId(), message.getKey(), message.operation().getStoreOpCode());
         encodedMsg.put(appendMessage.getPayload());
         return encodedMsg.array();
       case GET_AND_APPEND:
         GetAndAppendMessage getAndAppendMessage = (GetAndAppendMessage)message;
-        encodedMsg = ByteBuffer.allocate(MSG_TYPE_OFFSET + STORE_OP_CODE_OFFSET + CACHE_ID_LEN_OFFSET + KEY_OFFSET + cacheIdLen + getAndAppendMessage
+        encodedMsg = ByteBuffer.allocate(STORE_OP_CODE_SIZE + CACHE_ID_LEN_SIZE + KEY_SIZE + 2 * cacheIdLen + getAndAppendMessage
```

Comment:
this `STORE_OP_CODE_SIZE + CACHE_ID_LEN_SIZE + KEY_SIZE` is repeated for all messages - could be declared as a constant